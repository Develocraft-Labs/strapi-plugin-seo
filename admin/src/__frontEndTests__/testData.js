import trads from "../translations";

export const id = 1,
  firstname = "John",
  lastname = "Doe",
  email = "John.Doe@companyA.com",
  title = "title",
  created_by = {
    id,
    email,
    firstname,
    lastname,
  },
  created_at = "2021-10-12T11:49:01.042Z",
  updated_at = "2021-10-15T09:43:36.344Z",
  updated_by = { email, firstname, id, isActive: true, lastname },
  seo = {
    canonical: "https://www.companyA.com/articles",
    created_at: "2021-10-12T11:55:00.595Z",
    created_by: 1,
    id: 1,
    jsonLd: {},
    metaDescription: "Artists Desc",
    metaTitle: "Artists Title",
    ogImage: "Artist og image",
    ogTitle: "Artists Og Title",
    ogType: "Artists Og Type",
    ogUrl: "Artists og url",
    title: "Artists Seo",
    updated_at,
    updated_by: 1,
  },
  resourceData = {
    title,
    id,
    created_by,
  },
  collectionName = "Artists",
  pagination = { page: 1, total: 13, pageSize: 5, pageCount: 3 },
  uid = "application::artists.artists",
  fullResults = [
    { created_at, created_by, id, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 2, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 3, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 4, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 5, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 6, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 7, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 8, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 9, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 10, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 11, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 12, seo, title, updated_at, updated_by },
    { created_at, created_by, id: 13, seo, title, updated_at, updated_by },
  ],
  seoId = 1,
  apiID = "artists",
  attributes = {
    seo: {
      dominant: false,
      nature: "oneWay",
      plugin: "seo",
      private: false,
      target: "plugins::seo.seo",
      unique: false,
    },
    title,
  },
  schema = {
    attributes,
    collectionName,
    connection: "default",
    description: "",
    draftAndPublish: true,
    kind: "collectionType",
    name: "Artists",
    pluginOptions: {},
    restrictRelationsTo: null,
    visible: true,
  },
  resourceSchema = { collectionName, schema },
  pluginUID = "plugins::seo.seo",
  projectCollectionTypes = [{ apiID, schema, uid }],
  selectedSeo = {
    canonical: "https://www.companyA.com/articles",
    created_at: "2021-10-12T11:55:00.595Z",
    created_by: created_by,
    id: 1,
    jsonLd: {},
    metaDescription: "Artists Desc",
    metaTitle: "Artists Title",
    ogImage: "Artist og image",
    ogTitle: "Artists Og Title",
    ogType: "Artists Og Type",
    ogUrl: "Artists og url",
    title: "Artists Seo",
    updated_at,
    updated_by: updated_by,
  },
  refreshOptions = () => {},
  metaTitle = "meta title",
  metaDescription = "meta description",
  canonical = "canonical",
  handleDeleteSeo = () => {},
  handleEditSeo = () => {},
  defaultLocale = "en",
  userEnabledLocales = [
    {
      code: "en",
      created_at: "2021-12-20T09:03:53.724Z",
      id: 1,
      isDefault: true,
      name: "English (en)",
      updated_at: "2021-12-20T09:03:53.724Z",
    },
  ],
  seos = [seo, fullResults];

export const resourceComponentTestData = {
  resourceData,
  resourceSchema,
};

export const seoFormComponentTestData = {
  pluginUID,
  seoId,
  projectCollectionTypes,
  resource: resourceComponentTestData,
};

export const contentTypesListComponentTestData = [
  {
    collectionName,
    fullResults,
    pagination,
    uid,
  },
];

export const translationPickerComponentTestData = {
  translations: trads,
  handleTranslation: () => {},
  value: "",
};

export const contentTypesItemComponentTestData = {
  uid,
  fullResults,
  pagination,
  collectionName,
};
export const previewComponentTestData = {
  selectedSeo,
  pluginUID,
  refreshOptions,
  projectCollectionTypes,
  handleDeleteSeo,
};

export const tableComponentTestData = {
  seos: { collectionName, fullResults, uid, pagination },
  handleDeleteSeo,
  handleEditSeo,
  userEnabledLocales,
  uid,
  defaultLocale,
};
