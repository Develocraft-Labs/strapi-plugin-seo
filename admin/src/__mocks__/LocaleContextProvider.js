import React, { createContext, useContext } from "react";

export const mockContextValue = {
  isI18nPluginInstalled: true,
  userEnabledLocales: [
    {
      code: "en",
      created_at: "2022-03-04T07:56:18.881Z",
      id: 1,
      isDefault: true,
      name: "English (en)",
      updated_at: "2022-03-04T07:56:18.881Z",
    },
  ],
  defaultLocale: "en",
};

export const LocaleContext = createContext(mockContextValue);

export const useLocaleContext = () => useContext(LocaleContext);

const LocaleContextProvider = ({ children }) => {
  return (
    <LocaleContext.Provider value={mockContextValue}>
      {children}
    </LocaleContext.Provider>
  );
};

export default LocaleContextProvider;
