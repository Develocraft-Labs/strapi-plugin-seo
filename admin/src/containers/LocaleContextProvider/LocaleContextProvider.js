import React, { createContext, useContext, useEffect, useState } from "react";
import { LoadingIndicatorPage, request } from "strapi-helper-plugin";
import getDefaultLocale from "../../utils/getDefaultLocale";
import getPlugins from "../../utils/getPlugins";

export const LocaleContext = createContext();

export const useLocaleContext = () => useContext(LocaleContext);

const LocaleContextProvider = ({ children }) => {
  const [isI18nPluginInstalled, setIsI18nPluginInstalled] = useState(null);
  const [userEnabledLocales, setUserEnabledLocales] = useState(null);
  const [defaultLocale, setDefaultLocale] = useState(null);
  const [loaded, setLoaded] = useState(false);

  const contextValue = {
    isI18nPluginInstalled,
    userEnabledLocales,
    defaultLocale,
  };

  useEffect(() => {
    const fetchUserEnabledLocales = async (isI18nPluginAvailable) => {
      if (isI18nPluginAvailable) {
        const userEnabledLocales = await request("/i18n/locales");
        const defLocale = await getDefaultLocale(userEnabledLocales);
        setDefaultLocale(defLocale);
        setUserEnabledLocales(userEnabledLocales);
      }
    };

    const isI18nInstalled = async () => {
      const { isI18nPluginAvailable } = await getPlugins();
      setIsI18nPluginInstalled(isI18nPluginAvailable);
      await fetchUserEnabledLocales(isI18nPluginAvailable);
      setLoaded(true);
    };
    isI18nInstalled();
  }, []);

  // Ensure we checked the locale and are ready to render the ui/fetch data
  if (!loaded) {
    return <LoadingIndicatorPage />;
  }

  return (
    <LocaleContext.Provider value={contextValue}>
      {children}
    </LocaleContext.Provider>
  );
};

export default LocaleContextProvider;
