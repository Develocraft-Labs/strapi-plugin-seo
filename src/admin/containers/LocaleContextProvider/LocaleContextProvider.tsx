// @ts-ignore
import { request, LoadingIndicatorPage } from '@strapi/helper-plugin';
import React, {
  createContext,
  useContext,
  useEffect,
  useState,
  useMemo,
  FC,
} from 'react';

import getDefaultLocale from '../../utils/getDefaultLocale';
import getContentTypes from '../../utils/getContentTypes';
import { ILocale } from '../../interfaces/Locales';
import ROUTES from '../../utils/routes';
import { ILocaleContext } from '../../interfaces/LocaleContext';
import isI18nStrapiPluginInstalled from '../../utils/isI18nStrapiPluginInstalled';

export const LocaleContext = createContext<ILocaleContext>({
  isI18nPluginInstalled: false,
  userEnabledLocales: [],
  defaultLocale: '',
});

export const useLocaleContext = (): ILocaleContext => useContext(LocaleContext);

const LocalContextProvider: FC = ({ children }) => {
  const [isI18nPluginInstalled, setIsI18nPluginInstalled] = useState(false);
  const [userEnabledLocales, setUserEnabledLocales] = useState<ILocale[]>([]);
  const [defaultLocale, setDefaultLocale] = useState('');
  const [loaded, setLoaded] = useState(false);

  const contextValue = useMemo(
    () => ({
      isI18nPluginInstalled,
      userEnabledLocales,
      defaultLocale,
    }),
    [isI18nPluginInstalled, userEnabledLocales, defaultLocale]
  );

  useEffect(() => {
    const fetchUserEnabledLocales = async (isI18nPluginAvailable: boolean) => {
      if (isI18nPluginAvailable) {
        const userEnabledLocalesData: ILocale[] = await request(
          ROUTES.I18N_LOCALES
        );
        const defLocale = await getDefaultLocale(userEnabledLocalesData);

        setDefaultLocale(defLocale);
        setUserEnabledLocales(userEnabledLocalesData);
      }
    };

    const isI18nInstalled = async () => {
      const { plugins } = await getContentTypes();
      const isI18nPluginAvailable = isI18nStrapiPluginInstalled(plugins);

      setIsI18nPluginInstalled(isI18nPluginAvailable);
      await fetchUserEnabledLocales(isI18nPluginAvailable);
      setLoaded(true);
    };
    isI18nInstalled();
  }, []);

  // Ensure we checked the locale and are ready to render the ui/fetch data
  if (!loaded) {
    return <LoadingIndicatorPage />;
  }

  return (
    <LocaleContext.Provider value={contextValue}>
      {children}
    </LocaleContext.Provider>
  );
};

export default LocalContextProvider;
