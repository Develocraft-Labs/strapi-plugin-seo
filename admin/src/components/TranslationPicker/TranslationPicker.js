import React from "react";
import { Select } from "strapi-helper-plugin";

import createDropDownOptions from "../../utils/createDropDownOptions";
import styled from "styled-components";
import { useLocaleContext } from "../../containers/LocaleContextProvider/LocaleContextProvider";
import createTranslationObject from "../../utils/createTranslations";

const TranslationSelectContainer = styled.div`
  margin: 0rem 0rem 1.5rem 0rem;
  background-color: #ffffff;
  border-radius: 2px;
  min-width: 20%;
  max-width: 20%;
  align-self: flex-end;
`;

/**
 * Component that renders a dropdown with avaliable locales.
 * @returns {JSX.Element} Drop down.
 */
const TranslationPicker = ({ handleTranslation, value }) => {
  const { userEnabledLocales } = useLocaleContext();

  const translations = createTranslationObject(userEnabledLocales);

  if (!translations) return <option></option>;

  const availableTranslations = Object.keys(translations);
  const translationsLength = availableTranslations?.length;
  let options = [];

  if (translationsLength && translationsLength > 0 && userEnabledLocales)
    options = createDropDownOptions(availableTranslations, userEnabledLocales);

  return (
    <TranslationSelectContainer>
      <Select
        style={{
          maxWidth: "200px",
          fontSize: "1.3rem",
          alignSelf: "center",
        }}
        selectedValue={value}
        onChange={handleTranslation}
      >
        {options}
      </Select>
    </TranslationSelectContainer>
  );
};

export default TranslationPicker;
