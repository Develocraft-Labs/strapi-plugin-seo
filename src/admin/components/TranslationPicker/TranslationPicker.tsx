// @ts-ignore
import { Select, Option } from '@strapi/design-system/Select';
import React, { FC } from 'react';

import { useLocaleContext } from '../../containers/LocaleContextProvider/LocaleContextProvider';
import createTranslationHash from '../../utils/createTranslationHash';
import DropDownOptions from '../ui/DropDown/DropDownOptions';

interface ITranslationPicker {
  handleTranslation: (selectedValue: string) => void;
  value: string;
}

const TranslationPicker: FC<ITranslationPicker> = ({
  handleTranslation,
  value,
}) => {
  const { userEnabledLocales } = useLocaleContext();

  const translations = createTranslationHash(userEnabledLocales);
  if (!translations) return <Option />;

  const avaliableTranslations = Object.keys(translations);
  const translationsLength = avaliableTranslations?.length;

  if ((!translationsLength && translationsLength < 0) || !userEnabledLocales) {
    return <div />;
  }

  return (
    <Select value={value} onChange={handleTranslation}>
      {DropDownOptions({ avaliableTranslations, userEnabledLocales })}
    </Select>
  );
};

export default TranslationPicker;
