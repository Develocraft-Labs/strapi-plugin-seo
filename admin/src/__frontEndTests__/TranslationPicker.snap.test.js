/**
 * @jest-environment jsdom
 */
import React from "react";
import renderer from "react-test-renderer";

import TranslationPicker from "../components/TranslationPicker/TranslationPicker";
import LocaleContextProvider from "../__mocks__/LocaleContextProvider";
import { translationPickerComponentTestData } from "./testData";

// mock GlobalPagination component from strapi-helper-plugin
jest.mock("strapi-helper-plugin", () => {
  const Input = jest.fn((props) => (
    <input value={props.value} onChange={props.onChange} name={props.name} />
  ));
  const options = (props) => (
    <option value={props.title} id={props.id}>
      {props.title}
    </option>
  );
  const ReactSelect = jest.fn((props) => <Input props={props} />);
  const Select = jest.fn((props) => (
    <ReactSelect selectedValue={props.value} onChange={props.handleChange}>
      {options}
    </ReactSelect>
  ));

  return {
    LoadingIndicatorPage: () => null,
    Select: Select,
  };
});

jest.mock("../containers/LocaleContextProvider/LocaleContextProvider.js", () =>
  require("../__mocks__/LocaleContextProvider")
);

describe("SNAP TESTING TRANSLATION PICKER COMPONENT", () => {
  const { handleTranslation, translations, value } =
    translationPickerComponentTestData;

  it("Should Render Successfully", () => {
    const component = renderer.create(
      <LocaleContextProvider>
        <TranslationPicker
          handleTranslation={handleTranslation}
          translations={translations}
          value={value}
        />
      </LocaleContextProvider>
    );
    const tree = component.toJSON();
    expect(component).toBeDefined();
    expect(tree).toMatchSnapshot();
  });
});
