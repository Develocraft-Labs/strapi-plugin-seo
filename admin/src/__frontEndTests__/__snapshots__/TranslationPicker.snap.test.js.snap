// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAP TESTING TRANSLATION PICKER COMPONENT Should Render Successfully 1`] = `
.c0 {
  margin: 0rem 0rem 1.5rem 0rem;
  background-color: #ffffff;
  border-radius: 2px;
  min-width: 20%;
  max-width: 20%;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
}

<div
  className="c0"
>
  <input />
</div>
`;
